controllers/servicefence_controller.go

```diff
@@ -127,22 +148,143 @@ func (r *ServicefenceReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error
 			req.NamespacedName, rev, r.env.IstioRev())
 		return reconcile.Result{}, nil
 	}
-	log.Infof("get serviceFence, %+v", req.NamespacedName)
+	log.Infof("ServicefenceReconciler got serviceFence request, %+v", req.NamespacedName)
 
 	// 资源更新
 	diff := r.updateVisitedHostStatus(instance)
 	r.recordVisitor(instance, diff)
 	if instance.Spec.Enable {
-		if r.source != nil {
-			r.source.WatchAdd(types.NamespacedName{Name: instance.Name, Namespace: instance.Namespace})
+		r.interestMeta[req.NamespacedName.String()] = true
+		for k, v := range r.interestMeta {
+			r.InterestMetaCopy[k] = v
+		}
+		for k := range r.InterestMetaCopy {
+			if !r.interestMeta[k] {
+				delete(r.InterestMetaCopy, k)
+			}
 		}
 		err = r.refreshSidecar(instance)
 	}
 
 	return ctrl.Result{}, err
 }
 
+// call back function for watcher producer
+func (r *ServicefenceReconciler) handleWatcherEvent(event trigger.WatcherEvent) metric.QueryMap {
+	r.reconcileLock.RLock()
+	defer r.reconcileLock.RUnlock()
+
+	// check event
+	gvks := []schema.GroupVersionKind{
+		{Group: "networking.istio.io", Version: "v1beta1", Kind: "Sidecar"},
+	}
+	for _, gvk := range gvks {
+		if event.GVK == gvk && r.InterestMetaCopy[event.NN.String()] {
+			goto GenerateQueryMap
+		}
+	}
+	return nil
+
+	// generate query map for producer
+GenerateQueryMap:
+	qm := make(map[string][]metric.Handler)
+	var hs []metric.Handler
+	for name, pHandler := range r.env.Config.Metric.Prometheus.Handlers {
+		query := strings.ReplaceAll(pHandler.Query, "$namespace", event.NN.Namespace)
```

Comment:
we have two copies of this code block nearby

Reply:
replace by function generateHandler